module/os/freebsd/spl/spl_kstat.c

```diff
@@ -345,74 +349,54 @@ __kstat_create(const char *module, int instance, const char *name,
 	else
 		ksp->ks_data = kmem_zalloc(ksp->ks_data_size, KM_SLEEP);
 
-	/*
-	 * Some kstats use a module name like "zfs/poolname" to distinguish a
-	 * set of kstats belonging to a specific pool.  Split on '/' to add an
-	 * extra node for the pool name if needed.
-	 */
+	sysctl_ctx_init(&ksp->ks_sysctl_ctx);
+
 	(void) strlcpy(buf, module, KSTAT_STRLEN);
-	module = buf;
-	pool = strchr(module, '/');
-	if (pool != NULL)
-		*pool++ = '\0';
 
 	/*
-	 * Create sysctl tree for those statistics:
-	 *
-	 *	kstat.<module>[.<pool>].<class>.<name>
+	 * Walk over the module name, splitting on '/', and create the
+	 * intermediate nodes.
 	 */
-	sysctl_ctx_init(&ksp->ks_sysctl_ctx);
-	root = SYSCTL_ADD_NODE(&ksp->ks_sysctl_ctx,
-	    SYSCTL_STATIC_CHILDREN(_kstat), OID_AUTO, module, CTLFLAG_RW, 0,
-	    "");
-	if (root == NULL) {
-		printf("%s: Cannot create kstat.%s tree!\n", __func__, module);
-		sysctl_ctx_free(&ksp->ks_sysctl_ctx);
-		free(ksp, M_KSTAT);
-		return (NULL);
-	}
-	if (pool != NULL) {
-		root = SYSCTL_ADD_NODE(&ksp->ks_sysctl_ctx,
-		    SYSCTL_CHILDREN(root), OID_AUTO, pool, CTLFLAG_RW, 0, "");
+	root = NULL;
+	p = buf;
+	while ((frag = strsep(&p, "/")) != NULL) {
+		root = SYSCTL_ADD_NODE(&ksp->ks_sysctl_ctx, root ?
+		    SYSCTL_CHILDREN(root) : SYSCTL_STATIC_CHILDREN(_kstat),
+		    OID_AUTO, frag, CTLFLAG_RW, 0, "");
 		if (root == NULL) {
-			printf("%s: Cannot create kstat.%s.%s tree!\n",
-			    __func__, module, pool);
+			printf("%s: Cannot create kstat.%s tree!\n",
+			    __func__, buf);
 			sysctl_ctx_free(&ksp->ks_sysctl_ctx);
 			free(ksp, M_KSTAT);
 			return (NULL);
 		}
+		if (p != NULL && p > frag)
```

Comment:
Isn't `p > frag` always true here if it is not `NULL`?

Reply:
Yeah, good point. I think I was just being overly defensive. Will be removed in next push.